datacapture/src/main/java/com/google/android/fhir/datacapture/validation/ConstraintExtensionValidator.kt

```diff
@@ -0,0 +1,85 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.fhir.datacapture.validation
+
+import android.content.Context
+import com.google.android.fhir.datacapture.extensions.ConstraintSeverityTypes
+import com.google.android.fhir.datacapture.extensions.EXTENSION_QUESTIONNAIRE_CONSTRAINT_EXPRESSION
+import com.google.android.fhir.datacapture.extensions.EXTENSION_QUESTIONNAIRE_CONSTRAINT_HUMAN
+import com.google.android.fhir.datacapture.extensions.EXTENSION_QUESTIONNAIRE_CONSTRAINT_SEVERITY
+import com.google.android.fhir.datacapture.extensions.EXTENSION_QUESTIONNAIRE_CONSTRAINT_URL
+import com.google.android.fhir.datacapture.extensions.asStringValue
+import com.google.android.fhir.datacapture.fhirpath.ExpressionEvaluator
+import com.google.android.fhir.datacapture.fhirpath.evaluateToBoolean
+import org.hl7.fhir.r4.model.CodeType
+import org.hl7.fhir.r4.model.Expression
+import org.hl7.fhir.r4.model.Questionnaire.QuestionnaireItemComponent
+import org.hl7.fhir.r4.model.QuestionnaireResponse
+import org.hl7.fhir.r4.model.QuestionnaireResponse.QuestionnaireResponseItemComponent
+
+// TODO: Add constraint support for global case, add questionnaire in constructor, and process that
+// in the validate function, https://github.com/google/android-fhir/issues/2479
+internal class ConstraintExtensionValidator(
+  private val questionnaireResponse: QuestionnaireResponse,
+  private val expressionEvaluator: ExpressionEvaluator,
+) : QuestionnaireResponseItemConstraintValidator {
+  override suspend fun validate(
+    questionnaireItem: QuestionnaireItemComponent,
+    questionnaireResponseItem: QuestionnaireResponseItemComponent,
+    context: Context,
+  ): List<ConstraintValidator.Result> {
+    return questionnaireItem.extension
+      .filter { extension ->
+        // TODO: Add constraint support for warning case, update the [ConstraintValidator.Result]
+        // data class to also include warning state,
+        // https://github.com/google/android-fhir/issues/2480
+        extension.url == EXTENSION_QUESTIONNAIRE_CONSTRAINT_URL &&
+          (extension.getExtensionByUrl(EXTENSION_QUESTIONNAIRE_CONSTRAINT_SEVERITY).value
+              as CodeType)
+            .valueAsString == ConstraintSeverityTypes.ERROR.code
+      }
+      .map { extension ->
+        val expression =
+          Expression().apply {
+            language = "text/fhirpath"
+            expression =
+              extension
+                .getExtensionByUrl(EXTENSION_QUESTIONNAIRE_CONSTRAINT_EXPRESSION)
+                .value
+                .asStringValue()
+          }
+        val variables = expressionEvaluator.extractDependentVariables(expression, questionnaireItem)
+        val isValid =
+          evaluateToBoolean(
+            questionnaireResponse,
+            questionnaireResponseItem,
+            expression.expression,
+            variables,
+          )
```

Comment:
Can you use `expressionEvaluator.evaluateExpression` directly ?
```suggestion
val isValid = expressionEvaluator.evaluateExpression(
    questionnaireItem,
    questionnaireResponseItem,
    expression
).convertToBoolean()
```

I think there is a convertToBoolean function for List<Base> in `FhirPathUtil`.